Add Error Tracking Standalone Config option

Guillaume-Barrier opened the pull request:
### What does this PR do?
This PR adds ErrorTrackingStandalone config option as a boolean. When set to true, all samplers but the error sampler are bypassed. Only chunks that contain an **error span** or a **span with exception span events** are run through the error sampler. Kept spans are specifically tagged.

### Motivation
We want to offer users the opportunity to have Error Tracking Standalone, i.e. a way to gather backend errors with lower cost than buying APM - but with upsell in mind. ETBS only relies on chunks that contain errors, so only the error sampler should be run.

https://datadoghq.atlassian.net/browse/ERRORT-4747

### QA
The flag was tested locally with a dummy app, and later we deployed a custom agent build in the `porygon` experimental cluster in staging so that we could point one of our services (`errors-api`) to this version of the agent and have "real world" behaviour to validate behaviour.

Guillaume-Barrier committed:
Consider spans with exception spanEvents as errors

Error Tracking will support OpenTelemetry exception span events as
issues. The sampler should not drop spans that don't have an error
status but do have exception span events.

Guillaume-Barrier committed:
Add Error Tracking Standalone Config option

ErrorTrackingStandalone is a boolean. When set to true, all samplers but
the error sampler are bypassed. Kept spans are specifically tagged.

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/support-span-events-in-error-sampler

Guillaume-Barrier committed:
Add tag in receive span v2

There is now a second method to convert otlp spans to dd spans. Adding
the has_exception tag in this method as well.

Guillaume-Barrier committed:
Merge branch 'guillaume.barrier/support-span-events-in-error-sampler' into guillaume.barrier/add-error-tracking-standalone-config-option

Guillaume-Barrier committed:
Remove tag clean up

Removing the _dd.span_events.has_exception tag requires to loop through
the whole chunk which is not ideal. Also:
- tag is hidden (although visible in devtools)
- indication of why the chunk was sampled if there is no error span
- error sampler can be disabled so no clean up
And while it is there, this tag could also be used at processing to
detect exception span events without even unmarshalling.

Guillaume-Barrier committed:
Merge branch 'guillaume.barrier/support-span-events-in-error-sampler' into guillaume.barrier/add-error-tracking-standalone-config-option

Guillaume-Barrier committed:
Remove tag clean up

Removing the _dd.span_events.has_exception tag requires to loop through
the whole chunk which is not ideal. Also:
- tag is hidden (although visible in devtools)
- indication of why the chunk was sampled if there is no error span
- error sampler can be disabled so no clean up
And while it is there, this tag could also be used at processing to
detect exception span events without even unmarshalling.

Guillaume-Barrier committed:
Add missing setup of config ET standalone

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-error-tracking-standalone-config-option

Guillaume-Barrier committed:
Only consider exception span events for ETS

mackjmr commented on pkg/trace/api/otlp.go:
```diff
@@ -517,6 +517,12 @@ func (o *OTLPReceiver) convertSpan(rattr map[string]string, lib pcommon.Instrume
 	if in.Events().Len() > 0 {
 		transform.SetMetaOTLP(span, "events", transform.MarshalEvents(in.Events()))
 	}
+	for i := range in.Events().Len() {
```
Were already iterating through events/ checking for exceptions [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/transform/transform.go#L333), may be more efficient to add `span.Meta["_dd.span_events.has_exception"] = "true"` below that line.

It also seems like `Status2Error` gets called in both paths you changed, so moving to there would only require adding the logic in 1 place.

Guillaume-Barrier replied:
From what I can see, there is a check in `Status2Error` [here](https://github.com/DataDog/datadog-agent/blob/c079441bf0d6b2fc6eeaabb95d9fae69a9da4e34/pkg/trace/transform/transform.go#L325) so that it is only applied to error spans.

The point being to consider non error spans that do contain exception span events in the error sampler (in addition to error spans), I don't think I can move the tag setting there unfortunately.

mackjmr replied:
> The point being to consider non error spans that do contain exception span events in the error sampler

Ah, wasn't aware of this. Agreed you can't add to `Status2Error` then. May be good to still create a func to avoid duplication but is fine as is too, approving.

Guillaume-Barrier replied:
Moved into its own func in https://github.com/DataDog/datadog-agent/pull/30065/commits/7f8a8e5fc5efccf9bc22237990d5d91dc5616f30, cleaner indeed!

pgimalac commented on pkg/config/config_template.yaml:
```diff
@@ -1452,6 +1452,14 @@ api_key:
+  ## @param error_tracking_standalone - object - optional
+  ## Enables Error Tracking Standalone
+  ##
+  #error_tracking_standalone:
+  ## @env DD_APM_ERROR_TRACKING_STANDALONE_ENABLED - boolean - optional - default: false
+  ## Enables or disables Error Tracking Standalone
+  # enabled: false
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏You need to indent subkeys (see `internal_profiling.enabled` below for example)
```suggestion
  # error_tracking_standalone:
  #
    ## @param enabled - boolean - optional - default: false
    ## @env DD_APM_ERROR_TRACKING_STANDALONE_ENABLED - boolean - optional - default: false
    ## Enables or disables Error Tracking Standalone
    # enabled: false
```

Guillaume-Barrier replied:
Addressed in https://github.com/DataDog/datadog-agent/pull/30065/commits/f62c51deb0fcc62ab5f6aded049aaaa24ca862ee

Based the work on what was done for the Probabilistic Sampler, but yes it does look clearer this way :)

pgimalac commented on pkg/config/setup/apm.go:
```diff
@@ -112,6 +112,7 @@ func setupAPM(config pkgconfigmodel.Setup) {
 	config.BindEnv("apm_config.probabilistic_sampler.enabled", "DD_APM_PROBABILISTIC_SAMPLER_ENABLED")
 	config.BindEnv("apm_config.probabilistic_sampler.sampling_percentage", "DD_APM_PROBABILISTIC_SAMPLER_SAMPLING_PERCENTAGE")
 	config.BindEnv("apm_config.probabilistic_sampler.hash_seed", "DD_APM_PROBABILISTIC_SAMPLER_HASH_SEED")
+	config.BindEnv("apm_config.error_tracking_standalone.enabled", "DD_APM_ERROR_TRACKING_STANDALONE_ENABLED")
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏It is best to specify the default value, as using `config.Get(...)` would return `nil` otherwise for example
```suggestion
	config.BindEnvAndSetDefault("apm_config.error_tracking_standalone.enabled", false, "DD_APM_ERROR_TRACKING_STANDALONE_ENABLED")
```

Guillaume-Barrier replied:
Addressed in https://github.com/DataDog/datadog-agent/pull/30065/commits/f62c51deb0fcc62ab5f6aded049aaaa24ca862ee

pgimalac reviewed (approved):
LGTM for ASC, just a few nitpicks

Guillaume-Barrier committed:
Address comment

Indent subkeys in the config template and set default to false for error
tracking standalone flag.

Guillaume-Barrier committed:
Move span tagging into its own method

ichinaski commented on pkg/trace/agent/agent.go:
```diff
@@ -605,6 +612,48 @@ func isManualUserDrop(pt *traceutil.ProcessedTrace) bool {
+// Also sets "DroppedTrace" on the chunk.
+func (a *Agent) errorSampling(now time.Time, ts *info.TagStats, pt *traceutil.ProcessedTrace) (keep bool, numEvents int) {
+	sampled := a.runErrorSampler(now, *pt)
+	numEvents = len(a.getAnalyzedEvents(pt, ts))
```
do we really need analyzed spans when only this error sampler is enabled? 🤔

Guillaume-Barrier replied:
tbh I don't know a lot about analyzed spans but as it is set to `true` when running the error sampler in `runSamplers()`, I figured I would have it as well

Guillaume-Barrier replied:
Set `setAnalyzedSpans` bool to `false` in https://github.com/DataDog/datadog-agent/pull/30065/commits/96d9668e722c806ed325dc2ae8f6950e363eae39

ichinaski commented on pkg/trace/agent/agent.go:
```diff
@@ -605,6 +612,48 @@ func isManualUserDrop(pt *traceutil.ProcessedTrace) bool {
+	return false
+}
+
+func traceContainsErrorOrExceptionSpanEvent(trace pb.Trace) bool {
```
We have a very similar function already, maybe it's worth extending `traceContainsError()` adding a boolean parameter to determine whether exceptions should be considered?

Guillaume-Barrier replied:
that also works! done in https://github.com/DataDog/datadog-agent/pull/30065/commits/ce0d656b2e5348fe6bb7b3afd91c9cce238b89e0

Guillaume-Barrier commented on pkg/trace/agent/agent.go:
```diff
@@ -561,8 +561,15 @@ func (a *Agent) ProcessStats(in *pb.ClientStatsPayload, lang, tracerVersion stri
+	// Trace chunks that don't contain errors are dropped.
+	if a.conf.ErrorTrackingStandalone {
+		return a.errorSampling(now, ts, pt)
+	}
```
That may be more of a Product question but here is the reasoning:

- if a user sets a host to ETS, we only want to keep chunks with errors (or exception span events)
- the error sampler itself may not be needed - likely users would like all their errors and then use exclusion filters - but we can still leave it if they want to sample
- apart from the priority sampler with manual drop, all sampling decisions could be to keep regardless of the presence of errors - which we don't want as we only need to send chunks with errors

Then I have to admit I didn't know the probabilistic sampler talked to the tracer, but I guess we wouldn't need that if the host is set as ETS and we never run it?

I discussed with @dussault-antoine this week and we concluded that it was fine to bypass all samplers but the error sampler.

Then I don't remember if I considered putting the logic in `runSamplers()`, maybe that would give better readability?

ichinaski replied:
👍 From a product perspective, thanks for confirming that.

For readability and to keep things in the same place, I believe `runSamplers` will be a better place to decide which samplers to run, yes.

ajgajg1134 replied:
Unfortunately I think I need to answer this question with another question: If ErrorTrackingStandalone is disabled does that mean the customer is for sure NOT using APM and should not be able to view complete traces?
If so we should document this config value more explicitly with that in mind. There are a lot of features / functionality in the trace-agent so I worry that having this sort of "special mode" might accidentally infect a bunch of other features of the trace-agent (or that some thing will be missed and customers might accidentally enable or be billed for products unexpectedly) Is there an RFC or design doc I can review here?

Guillaume-Barrier replied:
Is it ok if it goes through SingleSpanSampling? if we don't keep a chunk, hopefully it is not kept

Guillaume-Barrier replied:
@ajgajg1134 ETS disabled is standard APM
ETS enabled acts as a filter on APM: the agent only sends chunks that contain errors

Here is the [product brief](https://docs.google.com/document/d/13ITxGJal0WO1zxjr3YDEdJkAGmlTmwam23cLFV3QhhA/edit?tab=t.0#heading=h.9hjfhrytnmop), and [another doc ](https://docs.google.com/document/d/1PoKs0D3CER3rL7jDlMEQGxkgMMKzkKibB44ygu6xFmc/edit?tab=t.0#heading=h.n0avhkm5m873) from @meghanlo

Guillaume-Barrier replied:
Moved sampling logic in `runSamplers()` in https://github.com/DataDog/datadog-agent/pull/30065/commits/96d9668e722c806ed325dc2ae8f6950e363eae39.

Also added a check in `sample()` to avoid single span sampling.

Guillaume-Barrier commented on pkg/trace/agent/agent.go:
```diff
@@ -605,6 +612,48 @@ func isManualUserDrop(pt *traceutil.ProcessedTrace) bool {
+			if span.Error != 0 || spanContainsExceptionSpanEvent(span) {
+				span.Meta["_dd.error_tracking_backend_standalone.error"] = "true"
+			} else {
+				span.Meta["_dd.error_tracking_backend_standalone.error"] = "false"
```
We are going to need the tag for billing, there is no other way we could know they come from ETS (we don't want to bill them for APM)

ichinaski replied:
You can have a single tag or flag at the `TraceChunk.Tags` level does represents the sampling mechanism (Standalone Error Sampling), and then, on the Intake, resolve whether the error/exception is found on any given span, then propagate the information as needed. Wouldn't this make the transport more efficient for both DD and customers?

Guillaume-Barrier replied:
Sounds good, tagged the chunk instead of each span in https://github.com/DataDog/datadog-agent/pull/30065/commits/96d9668e722c806ed325dc2ae8f6950e363eae39

git-thuerk-done reviewed (approved):
good for docs

Guillaume-Barrier committed:
Use traceContainsError to detect exception events

Guillaume-Barrier committed:
Move sampling logic in runSamplers()

Keep sampling things in the same place for readability. Also changes
tagging of spans - now only the chunk is tagged as
error_tracking_standalone. The checkAnalyticsEvents bool is also set to
false for ETS. Adds a ETS check to use or not the single span sampler.

ichinaski commented on pkg/trace/agent/agent.go:
```diff
@@ -623,13 +624,20 @@ func (a *Agent) getAnalyzedEvents(pt *traceutil.ProcessedTrace, ts *info.TagStat
 func (a *Agent) runSamplers(now time.Time, ts *info.TagStats, pt traceutil.ProcessedTrace) (keep bool, checkAnalyticsEvents bool) {
-	// run this early to make sure the signature gets counted by the RareSampler.
+	// ETS: chunks that don't contain errors (or spans with exception span events) are all dropped.
+	if a.conf.ErrorTrackingStandalone && traceContainsError(pt.TraceChunk.Spans, true) {
```
If `ErrorTrackingStandalone` is enabled, but no error is found on the trace, we will proceed to the other samplers logic. Not sure this is what we want, or what the documentation says? 🤔

Guillaume-Barrier replied:
No that's a mistake 😬 
Fixed in https://github.com/DataDog/datadog-agent/pull/30065/commits/68a8e33893890844f7fef95896c37baea22fe16a

Guillaume-Barrier committed:
Fix runSamplers for ETS

Return keep false if ETS enabled and no errors in chunk.

ichinaski commented on releasenotes/notes/add-error-tracking-standalone-config-option-1b88cedc9e164127.yaml:
```diff
@@ -0,0 +1,13 @@
+---
+features:
+  - |
+    Introducing the Error Tracking Standalone config option. Only span chunks
```
```suggestion
    APM: Introducing the Error Tracking Standalone config option. Only span chunks
```

Guillaume-Barrier committed:
Update releasenotes/notes/add-error-tracking-standalone-config-option-1b88cedc9e164127.yaml

Co-authored-by: Iñigo López de Heredia <inigo.lheredia@gmail.com>

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-error-tracking-standalone-config-option

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-error-tracking-standalone-config-option

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-error-tracking-standalone-config-option

Guillaume-Barrier committed:
Fix otlp_test

The test case was modified adding an exception span event, so the meta
had to be modified to add the has_exception tag.

Guillaume-Barrier committed:
Merge branch 'main' into guillaume.barrier/add-error-tracking-standalone-config-option

Guillaume-Barrier committed:
Fix config test

GianlucaBortoli commented:
/merge

GianlucaBortoli committed:
Merge branch 'main' into guillaume.barrier/add-error-tracking-standalone-config-option

GianlucaBortoli commented:
/merge

GianlucaBortoli committed:
Fix lint
